### PR TITLE
Add client setting to restore old freeze stars

### PR DESCRIPTION
**Pull request by sjrc6**
image

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Review by def- — APPROVED** [review]
Is it intentional that the freeze stars are always on a straight line? Looks wrong to me, and different than in your screenshot
Screenshot 2024-09-12 at 10 18 15

**Comment by sjrc6** [comment]
> Is it intentional that the freeze stars are always on a straight line? Looks wrong to me, and different than in your screenshot !

ah. I switched it from sin to std::sin to satisfy the build checks but that broke it :/

**Comment by sjrc6** [comment]
@def- try again now.

**Comment by def-** [comment]
Is it intentional that the freeze stars are always off to the side and not centered?

File: src/game/client/gameclient.cpp [hunk 1]
```diff
@@ -2071,6 +2071,33 @@ void CGameClient::OnNewSnapshot()
...
 
+	if(g_Config.m_ClFreezeStars)
+	{
+		for(auto &Character : m_Snap.m_aCharacters)
+		{
+			if(Character.m_Active && Character.m_HasExtendedData && Character.m_PrevExtendedData)
+			{
+				int FreezeTimeNow = Character.m_ExtendedData.m_FreezeEnd - Client()->GameTick(g_Config.m_ClDummy);
+				int FreezeTimePrev = Character.m_PrevExtendedData->m_FreezeEnd - Client()->PrevGameTick(g_Config.m_ClDummy);
+				vec2 Pos = vec2(Character.m_Cur.m_X, Character.m_Cur.m_Y);
```
Comment (Robyt3) [inline src/game/client/gameclient.cpp:2082]: Shouldn't this consider both the previous and the current positions and mix them?

**Comment by sjrc6** [comment]
> Is it intentional that the freeze stars are always off to the side and not centered?

Can you post a screenshot? They were always slightly off to the side.

File: src/game/client/gameclient.cpp (hunk 1 above)
Comment (sjrc6) [reply src/game/client/gameclient.cpp:2082]: no

**Comment by def-** [comment]
> Can you post a screenshot? They were always slightly off to the side.

Same as yours, ok

File: src/game/client/gameclient.cpp [hunk 2]
```diff
@@ -2071,6 +2071,33 @@ void CGameClient::OnNewSnapshot()
...
+				int StarsNow = (FreezeTimeNow + 1) / Client()->GameTickSpeed();
+				int StarsPrev = (FreezeTimePrev + 1) / Client()->GameTickSpeed();
+				if(StarsNow < StarsPrev || (StarsPrev == 0 && StarsNow > 0))
+				{
+					int Amount = StarsNow + 1;
+					float a = 3 * pi / 2;
+					float s = a - pi / 3;
+					float e = a + pi / 3;
```
[the remark this answers is not in the document]
Comment (Learath2) [reply src/game/client/gameclient.cpp:2090]: Or aleast `Mid`, `Min`, `Max`

File: src/game/client/gameclient.cpp [hunk 3]
```diff
@@ -2071,6 +2071,33 @@ void CGameClient::OnNewSnapshot()
...
+					for(int j = 0; j < Amount; j++)
+					{
+						float f = mix(s, e, (j + 1) / (float)(Amount + 2));
```
[the remark this answers is not in the document]
Comment (sjrc6) [reply src/game/client/gameclient.cpp]: sure

File: src/game/client/gameclient.cpp [hunk 4]
```diff
@@ -2071,6 +2071,33 @@ void CGameClient::OnNewSnapshot()
...
+						vec2 Dir = vec2(std::cos(f), std::sin(f));
+						m_Effects.DamageIndicator(Pos, Dir);
```
[the remark this answers is not in the document]
Comment (sjrc6) [reply src/game/client/gameclient.cpp]: sure

File: src/game/client/gameclient.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjrc6) [reply src/game/client/gameclient.cpp:2090]: > Or at least `Mid`, `Min`, `Max`

I like those

[the remark this answers is not in the document]
Comment (Robyt3) [reply src/game/client/gameclient.cpp]: > These are copied from the server code.

New code should use descriptive variable names. Feel free to also cleanup the server code in a separate commit.

File: src/game/client/gameclient.cpp (hunk 1 above)
Comment (Robyt3) [reply src/game/client/gameclient.cpp:2082]: Why?

**Comment by Robyt3** [comment]
When skipping long, short or only ticks in demos, lots of incorrect damage indicators keep being created now when in freeze. No damage indicators should be created while skipping in demos and all damage indicators (from guns) are properly deleted when skipping in demos.

**Comment by sjrc6** [comment]
> When skipping long, short or only ticks in demos, lots of incorrect damage indicators keep being created now when in freeze. No damage indicators should be created while skipping in demos and all damage indicators (from guns) are properly deleted when skipping in demos.

This is not fixable easily. The same bug applied to air jump effects. I think those are currently just removed from demos for this reason?

**Comment by Robyt3** [comment]
> > When skipping long, short or only ticks in demos, lots of incorrect damage indicators keep being created now when in freeze. No damage indicators should be created while skipping in demos and all damage indicators (from guns) are properly deleted when skipping in demos.
> 
> This is not fixable easily. The same bug applied to air jump effects. I think those are currently just removed from demos for this reason?

If it can't be fixed otherwise then add `&& !m_pClient->m_SuppressEvents` to the outer `if` condition so nothing is created while skipping in demos.

**Comment by sjrc6** [comment]
> If it can't be fixed otherwise then add `&& !m_pClient->m_SuppressEvents` to the outer `if` condition so nothing is created while skipping in demos.

ok

File: src/game/client/gameclient.cpp (hunk 1 above)
Comment (sjrc6) [reply src/game/client/gameclient.cpp:2082]: you're not entirely wrong but I think it should be this way to match the previous behavior.

File: src/game/client/gameclient.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjrc6) [reply src/game/client/gameclient.cpp]: > New code should use descriptive variable names. Feel free to also cleanup the server code in a separate commit.

Exact same code was accepted a month ago!
https://github.com/ddnet/ddnet/blob/66fb5d5d7f3b502bd4a3b1bef0d830690ce08c50/src/game/client/sixup_translate_snapshot.cpp#L426-L429

**Review by Robyt3 — APPROVED** [review]

**Comment by C0D3D3V** [comment]
Thank you ❤

**Pull request merged**